File: brutus.py
from mpi4py import MPI
import hashlib
import argparse
from tqdm import tqdm
import logging
from pyfiglet import Figlet
import colorlog
from termcolor import colored
from tabulate import tabulate
import time





comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()


def setup_logger(logger_name, rank):
    """Set up logger with color support."""
    logger = logging.getLogger(logger_name)
    logger.setLevel(logging.INFO)

    # Remove existing handlers
    for handler in logger.handlers[:]:
        logger.removeHandler(handler)

    # Choose color based on rank
    color = 'green' if rank == 0 else 'cyan'

    # Create a colored formatter
    formatter = colorlog.ColoredFormatter(
        f"%(log_color)s%(levelname)s:{logger_name}: %(message)s",
        log_colors={
            'DEBUG': 'cyan',
            'INFO': color,
            'WARNING': 'yellow',
            'ERROR': 'red',
            'CRITICAL': 'red,bg_white',
        },
        reset=True,
        style='%'
    )

    # Create a console handler and set its formatter to the colored formatter
    handler = logging.StreamHandler()
    handler.setFormatter(formatter)

    # Add the console handler to the logger
    logger.addHandler(handler)

    return logger


# Hashing function
def hash_password(password, hash_type):
    hash_func = hashlib.new(hash_type)
    hash_func.update(password.encode('utf-8'))
    return hash_func.hexdigest()



def read_password_list(path):
    with open(path, 'r') as file:
        dictionary = [line.strip() for line in file.readlines()]
    return dictionary



def countLines(path: str) -> int:
    with open(path, 'r') as file:
        return sum(1 for line in enumerate(file))

def splitChunks(lineCount: int, chunkSize: int) -> list:

    numChunks=lineCount//chunkSize
    chunks=[]

    for i in range(0, numChunks):
        start=i * chunkSize
        end=(i+1) * chunkSize
        if i == numChunks-1:
            end = lineCount
        chunks.append({'start': start, 'end': end})
    return chunks


def readChunk(path: str, start: int, end: int) -> list:
    chunk_data=[]
    with open(path, 'r') as file:
        for line_number,line in enumerate(file):
            if line_number > end:
                break
            if start <= line_number < end:
                chunk_data.append(line.strip())
    return chunk_data


def processChunk(chunk: list, password: str, hash_type: str) -> str:
    # hashed_password = hash_password(password, hash_type)
    for test_password in chunk:
        if hash_password(test_password, hash_type) == password:
            return test_password
    return None

def ascii_banner():

    f = Figlet(font='univers')
    print(colored(f.renderText('BRUTUS'), 'red'))
    f_small = Figlet(font='digital')
    print(colored(f_small.renderText('A simple password cracker using MPI'), 'red'))

def parameter_table(password, hash_type, dict_file, size,lines,chunkSize):
    # Create a table for the initial logs
    table = [
        ["Password Hash", password],
        ["Hashtype", hash_type],
        ["Dictionary File", dict_file],
        ["Number of Processes", size],
        ["Total Passwords to Try", lines],
        ["ChunkSize", chunkSize]
    ]
    table_str = tabulate(table, headers=["Parameter", "Value"], tablefmt="pipe")

    # Add table boundaries and change the table color
    table_str = colored(table_str, 'green')

    # Add top and bottom borders
    border = colored('*' * (len(table_str.split('\n')[0])-5), 'green')
    table_str = border + '\n' + table_str + '\n' + border + '\n\n'

    print(table_str)

def brute_force(dict_file,password, hash_type,chunkSize):

    
    start_time = time.time()
    logger_name = 'MASTER' if rank == 0 else f'SLAVE:{rank}'
    logger = setup_logger(logger_name, rank)


    if rank==0:
        ascii_banner()
        logger.info("Starting the brute force attack with Parameters")





        # chunkSize=10000
        final_result=(None,None)
    #   Send Password and hash_type to Slave Processes
        for i in range(1, size):
            comm.send(password, dest=i,tag=1)
            comm.send(hash_type, dest=i,tag=2)

        lines=countLines(dict_file)
    #     # !Modify this later to divide the chunks evenly
        if(lines>0 and lines<=chunkSize):
            chunkSize=lines//(size-1)


            
        
        parameter_table(password, hash_type, dict_file, size,lines,chunkSize)
        # proceed = input("Do you want to proceed? (yes/no): ")
        # if proceed.lower() != "yes":
        #     return
        
        # logger.info(f"Total Passwords to Try: {lines}")
        # logger.info(f"ChunkSize: {chunkSize}")


        current_chunk=0
        # * Sending first chunk to all processes
        chunks_queue=splitChunks(lines,chunkSize)
        for i in range(1, size):
            obj=chunks_queue[current_chunk]
            # logger.info(f"Chunk Starting Line: {obj['start']},Chunk Ending Line: {obj['end']}")
            chunk_data=readChunk(dict_file,obj['start'],obj['end'])
            current_chunk+=1
            comm.send(chunk_data, dest=i,tag=3)

        # * Getting Results from Slave Processes and Distributing the next chunk
        status = MPI.Status()
        while current_chunk < len(chunks_queue):
            logger.info(f"Current Chunk: {current_chunk} out of {len(chunks_queue)}")
            result = comm.recv(source=MPI.ANY_SOURCE, tag=10,status=status)
            slave_rank = status.Get_source()
            if result[0]:
                logger.info(f"Password found: {result[1]} by Rank {slave_rank}")
                final_result = result
                # Broadcast the termination signal to all slave processes
                for i in range(1, size):
                    comm.send(None, dest=i, tag=13)
                break
            else:
                # print ("TESTTTTT")
                obj=chunks_queue[current_chunk]
                logger.info (f"Sending next chunk to Rank {slave_rank}. {obj['start']}: {obj['end']}")
                chunk_data=readChunk(dict_file,obj['start'],obj['end'])
                comm.send(chunk_data, dest=slave_rank,tag=3)
                current_chunk+=1
                # !TODO: Send the next chunk to the process that just finished

       


        # * Collect any remaining results. 
        for i in range(1, size):
            # print ("YESSS")
            result = comm.recv(source=MPI.ANY_SOURCE, tag=10)
            if result[0]:  # If the password was found
                logger.info(f"Password found by worker {i}: {result[1]}, Terminating all processes.")
                final_result = result
                for i in range(1, size):
                    comm.send(None, dest=i, tag=99)  # Sending termination signal
                break
            else:
                logger.info(f"Rank {i} did not find the password.")
        

        for i in range(1, size):
            comm.send(None, dest=i, tag=99)
        
        comm.Barrier()
        end_time = time.time()
        time_taken = end_time - start_time
        if final_result[0]:
            

            logger.info("Finishing Execution")
            table = [
                ["Password", final_result[1]],
                ["Time Taken", time_taken]
            ]
            table_str = tabulate(table, headers=["Parameter", "Value"], tablefmt="pipe")
            table_str = colored(table_str, 'green')
            border = colored('*' * (len(table_str.split('\n')[0])-5), 'green')
            table_str = '\n\n'+border + '\n' + table_str + '\n' + border + '\n\n'
            print(table_str)
        else:


            logger.info("Finishing Execution")
            table = [
                ["Password", "Not Found"],
                ["Time Taken", time_taken]
            ]
            table_str = tabulate(table, headers=["Parameter", "Value"], tablefmt="pipe")
            table_str = colored(table_str, 'red')
            border = colored('*' * (len(table_str.split('\n')[0])-5), 'red')
            table_str = '\n\n'+border + '\n' + table_str + '\n' + border + '\n\n'
            print(table_str)
    
        
    else:
        # * Slave Process

        password=comm.recv(source=0,tag=1)
        hash_type=comm.recv(source=0,tag=2)

        prcoessedChunk=1
        while True:
            chunk = comm.recv(source=0, tag=MPI.ANY_TAG)
            if chunk is None:
                logger.info("Termination Signal Recieved")
                break   
        
            
            result=processChunk(chunk,password,hash_type)
            logger.info(f"Processed Chunks: {prcoessedChunk}")
            prcoessedChunk+=1
            if result:
                logger.info(f"Password Found: {result}")
                comm.send((True,result), dest=0, tag=10)
                break
            else:
                logger.info("Password not found. Requesting next chunk")
                comm.send((False,None), dest=0, tag=10)

        # print (f"{rank} out from loop")
        comm.send((None,None),dest=0, tag=10)
        comm.Barrier()
        # print (f"{rank} sent termination signal")


    








if __name__=='__main__':

    
    parser = argparse.ArgumentParser(description="Brutus - A simple password cracker")
    parser.add_argument('--password', type=str, help="The password HASH to crack")
    parser.add_argument('--algorithm', type=str, choices=['md5', 'sha1', 'sha256'], default='sha256', help="Hash algorithm to use")
    parser.add_argument('--path', type=str, help="Path to the password list")
    parser.add_argument('--chunksize', type=int, help="Size of the chunk to be processed by each process")
    args = parser.parse_args()

    # path='./PasswordLists/10-million-password-list-top-1000.txt'
    brute_force(args.path,args.password,args.algorithm,args.chunksize)



    

File: brutus.go
package main

import (
	"bufio"
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha256"
	"fmt"
	"hash"
	"os"
	"strings"

	"github.com/emer/empi/v2/mpi"
)

type Chunk struct {
	start int
	end   int
}

func send_string(comm *mpi.Comm, rank int, tag int, str string) {
	strBytes := []byte(str)
	strLen := len(strBytes)
	arr := make([]int, 1)
	arr[0] = strLen
	comm.SendInt(rank, tag, arr)
	comm.SendU8(rank, tag, strBytes)
}

func recv_string(comm *mpi.Comm, rank int, tag int) string {
	arr := make([]int, 1)
	comm.RecvInt(rank, tag, arr)
	strBytes := make([]byte, arr[0])
	comm.RecvU8(rank, tag, strBytes)
	return string(strBytes)
}

func send_string_array(comm *mpi.Comm, rank int, tag int, strs []string) {
	// First, send the length of the array
	arrLen := len(strs)
	arr := make([]int, 1)
	arr[0] = arrLen
	comm.SendInt(rank, tag, arr)

	// Then, send each string in the array
	for _, str := range strs {
		send_string(comm, rank, tag, str)
	}
}

func recv_string_array(comm *mpi.Comm, rank int, tag int) []string {
	// First, receive the length of the array
	arr := make([]int, 1)
	comm.RecvInt(rank, tag, arr)
	arrLen := arr[0]

	// Then, receive each string in the array
	strs := make([]string, arrLen)
	for i := 0; i < arrLen; i++ {
		strs[i] = recv_string(comm, rank, tag)
	}

	return strs
}

func hashPassword(password string, hashType string) (string, error) {

	var h hash.Hash
	switch hashType {
	case "md5":
		h = md5.New()
	case "sha1":
		h = sha1.New()
	case "sha256":
		h = sha256.New()
	default:
		return "", fmt.Errorf("invalid hash type %s", hashType)
	}

	h.Write([]byte(password))
	bs := h.Sum(nil)
	// fmt.Println(password)
	// fmt.Printf("%x\n", bs)

	return fmt.Sprintf("%x", bs), nil
}

func countLines(path string) (int, error) {
	fmt.Println("Counting lines in file: ", path)
	file, err := os.Open(path)
	if err != nil {
		return 0, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	lineCount := 0
	for scanner.Scan() {
		lineCount++
	}

	if scanner.Err() != nil {
		return 0, scanner.Err()
	}

	return lineCount, nil
}

func splitChunks(lineCount int, size int) []Chunk {

	// Naive Implementation. All processes except the root process will get the same number of lines.
	chunkSize := lineCount / size
	chunks := make([]Chunk, size)
	fmt.Println("Chunk Size: ", size)

	for i := 1; i <= size; i++ {
		start := (i - 1) * chunkSize
		end := i * chunkSize
		if i == size {
			end = lineCount
		}
		chunks[i-1] = Chunk{start, end}
		fmt.Println("Start: ", start, " End: ", end)
		//go processChunk(path, start, end)
	}
	return chunks
}

func readChunk(path string, chunk Chunk) ([]string, error) {
	file, err := os.Open(path)
	if err != nil {
		fmt.Println("Error:", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	lineCount := 0
	chunkLines := make([]string, 0)
	for scanner.Scan() {
		if lineCount >= chunk.start && lineCount < chunk.end {
			line := scanner.Text()
			line = strings.TrimSpace(line) // remove newline character
			chunkLines = append(chunkLines, line)
		}
		lineCount++
	}
	return chunkLines, nil
}

func processChunk(password string, path string, chunk []string, hashType string) string{


	password_hash, _ := hashPassword(password, hashType)
	for _, line := range chunk {

		hashed, _ := hashPassword(line, hashType)

		// fmt.Println("Line Number ", lineCount, ":", line)
		// fmt.Println("Hashed Password: ", hashed)

		if hashed == password_hash {
			fmt.Println("Password Found: ", line)
			return line
		}
	}
	return ""
}


func main() {

	/* !TODO: Add the functionality for other workers to stop if password found by one  */

	// Initialize MPI
	mpi.Init()
	defer mpi.Finalize()

	rank := mpi.WorldRank()
	size := mpi.WorldSize()
	// Send byte slice
	comm, _ := mpi.NewComm(nil)

	fmt.Println("TEST:", size, rank)

	path := "./PasswordLists/10-million-password-list-top-1000.txt"
	// path := "./PasswordLists/weakpass_3w"

	if rank == 0 {

		chunkSize := 1000

		// Sending Passowrd to all processes
		password := "music"
		for i := 1; i < size; i++ {
			send_string(comm, i, 1, password)
		}

		// Counting the Lines in the Dictionary file
		lineCount, err := countLines(path)
		if err != nil {
			fmt.Println("Error:", err)
		} else {
			fmt.Println("Number of lines:", lineCount)
		}

		if lineCount <= chunkSize {
			chunkSize = size - 1
		}

		currentChunk := 0
		chunks_queue := splitChunks(lineCount, chunkSize)

		// Initially Distribute one chunk to each process
		for i := 1; i < size; i++ {
			chunk, err := readChunk(path, chunks_queue[currentChunk])
			if err != nil {
				fmt.Println("Error in reading Chunk:", chunks_queue[currentChunk], err)
			}
			currentChunk++
			send_string_array(comm, i, 1, chunk)
		}

		// Then send the remaining chunks to the processes as they finish

		// for i := 1; i < size; i++ {
		// 	send_string_array(comm, i, 1, arr[:])
		// }

		// lineCount, err := countLines(path)
		// if err != nil {
		// 	fmt.Println("Error:", err)
		// } else {
		// 	fmt.Println("Number of lines:", lineCount)
		// }
		// chunks := splitChunks(lineCount, size)
		// fmt.Println(chunks)

		// data := make([]int, 2)
		// for i := 1; i < size; i++ {

		// 	data[0] = chunks[i-1].start
		// 	data[1] = chunks[i-1].end
		// 	fmt.Println("Sending Chunk: ", data)
		// 	comm.SendInt(i, 10, data)
		// }

	} else {

		// Recieving Password to Crack
		password := recv_string(comm, 0, 1)
		println("Received Passowrd: ", password)

		// Recieving Chunk
		chunks := recv_string_array(comm, 0, 1)
		fmt.Println("Received Chunk: ", len(chunks))

		chunk_result:=processChunk(password, path, chunks, "md5")
		if chunk_result!=""{
			fmt.Println("Password Found by Rank: ", rank, "Password: ", chunk_result)
		}

		
		// fmt.Println("Received Array: ", len(strs))

		// fmt.Println("Slave Rank: ", rank, "Received Array: ", len(chunks))

		// data := make([]int, 2)
		// comm.RecvInt(0, 10, data)

		// processChunk(password, path, Chunk{data[0], data[1]}, "md5")
		// fmt.Println("Rank ", rank, "Received Chunk: ", data[0], data[1])
	}
}
